### PR TITLE
[Fixes  #7355] Introduce the adoption of groups to define users permissions, and define add_resource permission for users

### LINKED ISSUE #7355: Introduce the adoption of groups to define users permissions, and define add_resource permission for users
We want to improve the way permissions are managed in GeoNode, to make room for more advanced combinations of roles and permissions. The idea is to increase the usage of permissions to granularly drive the functionalities both in the backend and the client, sintead of simply relying on the "AnonymousUser" and the admin concepts.

We start by introducing the concpet of a "contributors" role (Django Group) which is assigned the base_addresourcebase. Every user will be assigned to the group by default, so that the current assumption that any registerd user can create resources is maintained.

The first usage of this permissions will be:

 - the client uses it to decide if the "Create ..." and "Save as..." should be visible for the user
 - replace the READ_ONLY checks in the backend to decide if resources can be created (the only case where they can't is when the read only mode is active)

More use cases to come, and more permissions will be added to the group and security checks will be done on top of these permissions.

## Expected Behavior
-  A `contributors` Django group is created and the `base_addresourcebase` permission is assigned to it. The creation of this group could follow the same approach that we're using to create (and assign) a user to the `anonymous` group, i.e. it's created the very first time we need it.
- Every user will be assigned, by default, to the `contributors` group.(except anonymous and admins)
- A new `People.perms` property will return all the set (unique list) of permissions from the user's groups, i.e. permissions will be the combination of all the group's (non-repeated) permissions the user is a member of. 

The returned permissions are a mapping from Django permissions to a new set of GeoNode constants that will be declared inside the `security` module.
The first constant will be:

`base_addresourcebase -> add_resource`

The same will be done along the way to convert all the GeoNode code to use constanst instead of the original Django / Guardian permission strings.

## Actual Behavior
- There is no contributor group and `base_addresorcebase`

## Steps to Reproduce the Problem

  1. New addition

## Specifications

  - GeoNode version: master
  - Installation method (manual, GeoNode Docker, SPCGeoNode Docker):
  - Platform:
  - Additional details:

### PR DESCRIPTION
**Pull request by marthamareal**
References: #7355

- Creates a new user group `contributors` with permission `base_addresourcebase`.
- Updates existing users to belong to a new contributors group.
- Includes user's group perms to `get_user_perms` function

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [x] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by giohappy** [comment]
@marthamareal what's the purpose of adding the base_addresourcebase permission to the Profile model?

**Comment by marthamareal** [comment]
> @marthamareal what's the purpose of adding the base_addresourcebase permission to the Profile model?

This was before we introduced using contributors group, let me change and create the permission on the group

File: geonode/security/constants.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Permissions constants
+
+GENERAL_PERMISSIONS = {
```
Comment (giohappy) [inline geonode/security/constants.py:3]: @maabdelghaffar I would rename this to `PERMISSIONS`, simply

File: geonode/security/models.py [hunk 2]
```diff
@@ -396,6 +399,10 @@ def get_user_perms(self, user):
         """
         Returns a list of permissions a user has on a given resource
         """
+        # To avoid circular import
+        from geonode.base.models import Configuration
+
+        config = Configuration.load()
         ctype = ContentType.objects.get_for_model(self)
         PERMISSIONS_TO_FETCH = VIEW_PERMISSIONS + ADMIN_PERMISSIONS + LAYER_ADMIN_PERMISSIONS
```
Comment (giohappy) [inline geonode/security/models.py:407]: @marthamareal I would also move VIEW_PERMISSIONS, ADMIN_PERMISSIONS, LAYER_ADMIN_PERMISSIONS into the permissions module.

Comment (marthamareal) [reply geonode/security/models.py:395]: Done

**Review by giohappy — DISMISSED** [review]

**Review by giohappy — CHANGES_REQUESTED** [review]
@marthamareal instead of using a command to assign the contributors group to users it would be better to use a migration. This way it runs automatically and only once.

Something like this would be fine:

```
# Assign the contributors group to users according to #7364

from django.db import migrations, models
from django.db.migrations.operations import RunPython

def forwards_func(apps, schema_editor):
    # assign contributors group to users
    pass

def reverse_func(apps, schema_editor):
    # remove contributors group from users
    pass

class Migration(migrations.Migration):

    dependencies = [
        ('people', '0031_merge_20210205_0824'),
    ]

    operations = [
        RunPython(forwards_func, reverse_func)
    ]

```

**Comment by marthamareal** [comment]
> @marthamareal instead of using a command to assign the contributors group to users it would be better to use a migration. This way it runs automatically and only once.
> 
> Something like this would be fine:
> 
> ```
> # Assign the contributors group to users according to #7364
> 
> from django.db import migrations, models
> from django.db.migrations.operations import RunPython
> 
> def forwards_func(apps, schema_editor):
>     # assign contributors group to users
>     pass
> 
> def reverse_func(apps, schema_editor):
>     # remove contributors group from users
>     pass
> 
> class Migration(migrations.Migration):
> 
>     dependencies = [
>         ('people', '0031_merge_20210205_0824'),
>     ]
> 
>     operations = [
>         RunPython(forwards_func, reverse_func)
>     ]
> ```

updated this

**Comment by afabiani** [comment]
@marthamareal LGTM complains about a new unused import introduced with this PR, please, take a look. Thanks.

**Review by giohappy — APPROVED** [review]

**Review by giohappy — CHANGES_REQUESTED** [review]
@marthamareal let's use lowercase for GeoNode permission names, so please change "ADD_RESOURCE" to "add_resource".

**Pull request merged**